modules/calib3d/src/rhorefc.cpp

```diff
@@ -0,0 +1,2426 @@
+}
+
+
+
+/**
+ * Allocate memory aligned to a boundary of MEMALIGN.
+ */
+
+static inline void*  almalloc(size_t nBytes){
+    if(nBytes){
+        unsigned char* ptr = (unsigned char*)malloc(MEM_ALIGN + nBytes);
+        if(ptr){
+            unsigned char* adj = (unsigned char*)(((intptr_t)(ptr+MEM_ALIGN))&((intptr_t)(-MEM_ALIGN)));
+            ptrdiff_t diff = adj - ptr;
+            adj[-1] = (unsigned char)(diff - 1);
+            return adj;
+        }
+    }
+
+    return NULL;
+}
+
+/**
+ * Free aligned memory.
+ *
+ * If argument is NULL, do nothing in accordance with free() semantics.
+ */
+
+static inline void   alfree(void* ptr){
+    if(ptr){
+        unsigned char* cptr = (unsigned char*)ptr;
+        free(cptr - (ptrdiff_t)cptr[-1] - 1);
+    }
+}
+
+/**
+ * Constructor for RHO_HEST_REFC.
+ *
+ * Does nothing. True initialization is done by initialize().
+ */
+
+RHO_HEST_REFC::RHO_HEST_REFC() : init(0){
+
+}
+
+/**
+ * Private copy constructor for RHO_HEST_REFC. Disabled.
+ */
+
+RHO_HEST_REFC::RHO_HEST_REFC(const RHO_HEST_REFC&) : init(0){
+
+}
+
+/**
+ * Destructor for RHO_HEST_REFC.
+ */
+
+RHO_HEST_REFC::~RHO_HEST_REFC(){
+    if(init){
+        finalize();
+    }
+}
+
+
+
+/**
+ * Initialize the estimator context, by allocating the aligned buffers
+ * internally needed.
+ *
+ * Currently there are 5 per-estimator buffers:
+ * - The buffer of m indexes representing a sample
+ * - The buffer of 16 floats representing m matches (x,y) -> (X,Y).
+ * - The buffer for the current homography
+ * - The buffer for the best-so-far homography
+ * - Optionally, the non-randomness criterion table
+ *
+ * Returns 0 if unsuccessful and non-0 otherwise.
+ */
+
+inline int    RHO_HEST_REFC::initialize(void){
+    ctrl.smpl   = (unsigned*)almalloc(SMPL_SIZE*sizeof(*ctrl.smpl));
+
+    curr.pkdPts = (float*)   almalloc(SMPL_SIZE*2*2*sizeof(*curr.pkdPts));
+    curr.H      = (float*)   almalloc(HSIZE);
+    curr.inl    = NULL;
+    curr.numInl = 0;
+
+    best.H      = (float*)   almalloc(HSIZE);
+    best.inl    = NULL;
+    best.numInl = 0;
+
+    nr.size     = 0;
+    nr.beta     = 0.0;
+
+    lm.ws       = NULL;
+    lm.JtJ      = NULL;
+    lm.tmp1     = NULL;
+    lm.Jte      = NULL;
+
+#ifdef _WIN32
+    fastSeed(rand());
+#else
+    fastSeed(random());
+#endif
+
+
+    int areAllAllocsSuccessful = ctrl.smpl   &&
+                                 curr.H      &&
+                                 best.H      &&
+                                 curr.pkdPts;
+
+    if(!areAllAllocsSuccessful){
+        finalize();
+    }else{
+        init = 1;
+    }
+
+    return areAllAllocsSuccessful;
+}
+
+/**
+ * Finalize.
+ *
+ * Finalize the estimator context, by freeing the aligned buffers used
+ * internally.
+ */
+
+inline void   RHO_HEST_REFC::finalize(void){
+    if(init){
+        alfree(ctrl.smpl);
+        alfree(curr.H);
+        alfree(best.H);
+        alfree(curr.pkdPts);
+
+        memset(this, 0, sizeof(*this));
```

Comment:
this is bad practice, this may include pointers to virtual tables etc. do not use memset() with this